[박민서] 4주차 과제 제출

m2nsp opened the pull request:
- member CRUD 구현

<!--
제목 예시: [이름] n주차 과제 제출
-->

## 📌 구현 기능  
- [x] member CRUD 구현

## 🗂️ 과제 정리  
- member CRUD 구현

- 회원가입
![스크린샷 2025-04-05 002535](https://github.com/user-attachments/assets/2fdc8495-09af-40ce-be74-e60741c6c12b)

- 멤버조회
![스크린샷 2025-04-05 002624](https://github.com/user-attachments/assets/51223783-2fc8-462f-8e6a-250981706b86)

- 회원 정보 수정
![스크린샷 2025-04-05 003633](https://github.com/user-attachments/assets/fe1b3900-41cf-49b5-a07e-7b4538d43286)

- 논리적 회원탈퇴
![스크린샷 2025-04-05 003712](https://github.com/user-attachments/assets/ad85d591-6dae-4353-8634-0cce6ed05810)

## ⚠️ 어려웠던 점  
- 스프링은 처음 해보는 거라 시간이 부족해서 각 상태 메세지 별 코드를 다 완벽하게 구현하지 못한 것 같습니다
- 에러 처리 방식이 아직 잘 와닿지 않는 것 같습니다.
- DTO에서의 getter와 entity에서의 getter의 차이가 아직 헷갈리는 것 같습니다.
- 
## 기타
- 아직 의존성 주입 개념에 대해서 정확하게 이해하지 못한 것 같아서 롬북을 사용하지 않고 구현해 보았습니다.
궁금한점
- 닉네임은 중복이 가능하게 해도 되는 건지
- 회원 조회 시 status가 inactive 인 사용자도 조회할 수 있도록 되는 건지

m2nsp committed:
✨feat: 4주차 과제

- member CRUD 구현

m2nsp committed:
✨feat: 4주차 과제

jettieb reviewed (approved):
민서님! 과제 수고많으셨어요ㅎㅎ 😍 
질문 주신것 답변 여기에 바로 드릴게요!
1. 에러처리의 경우 이번 세미나에서 쓴 것 처럼 `IllegalArgumentException`으로 진행하면 지금은 일단 서버에러(500 Internal Server Error)으로 나오게 될 거에요! 에러 처리는 차차 세미나에서 다룰 것 같습니다 😺
2. Entity의 getter의 경우는 db에 저장되어있는 엔티티 값을 가져올 때 사용하고, dto의 getter의 경우에는 api 응답 또는 요청 데이터 전달용입니다!
3. 닉네임 중복의 경우 커뮤니티마다 다를 것 같긴 한데, 제가 확인해보니까 에타는 중복 허용이네요!ㅎㅎ
4. 회원 조회 시 원래는 status가 inactive인 사용자는 제외하고 조회하는 게 맞을 것 같네요!! 근데 전체 사용자가 아닌 특정 사용자만 조회하기 위해서는 jpa 스터디 때 배웠던 jpql 혹은 메서드 쿼리를 사용해야 해요! 이것도 곧 세미나에서 다룰 것 같아요! 우선은 전체 조회하는 걸로 해보아요! 😄

하단에 코드리뷰 드릴게요!

jettieb commented on src/main/java/efub/assignment/community/member/MemberRepository.java:
보통 repository랑 service, controller도 패키지로 한 번 묶어서 만듭니다! domain이나 dto처럼 repository, service, controller 패키지도 만들어서 분리해주면 좋을 것 같아요:)

jettieb commented on src/main/java/efub/assignment/community/member/MemberService.java:
```diff
@@ -0,0 +1,115 @@
+        //MemberResponseDTO로 변환
+        return new MemberResponseDTO(
+                savedMember.getMemberId(),
+                savedMember.getStudentNumber(),
```
결과적으로는 동일하긴 하지만, 저는 개인적으로 dto로 변환하는 부분은 dto 클래스에서 미리 메소드로 정의하는걸 선호합니다! 
MemberResponseDTO클래스에서
```
public static MemberResponseDTO from(Member member) {
    return new MemberResponseDTO(
        member.getMemberId(),
        member.getStudentNumber(),
        member.getNickname(),
        member.getSchool(),
        member.getEmail()
    );
}
```
이걸 미리 메소드로 만들어두고, service에서는 
```
Member savedMember = memberRepository.save(newMember);
return MemberResponseDTO.from(savedMember);
```

이렇게 깔끔해지도록 저는 주로 사용합니다!

jettieb commented on src/main/java/efub/assignment/community/member/MemberService.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    //회원 조회
+    public MemberResponseDTO getMemberByMemberId(Long memberId){
```
조회 메서드에는 `@Transactional(readOnly = true)`를 붙이면 성능 최적화와 의도 명시에 도움이 됩니다!

jettieb commented on src/main/java/efub/assignment/community/member/MemberService.java:
```diff
@@ -0,0 +1,115 @@
+        return new MemberResponseDTO(
+                member.getMemberId(),
+                member.getStudentNumber(),
+                member.getNickname(),
```
여기도 위와 마찬가지로 MemerResponseDTO에 미리 member를 dto로 만들어주는 메소드를 만들어준 후에,
`return new MemberResponseDTO.from(member);`로 깔끔하게 만들어주는게 좋을 것 같아용! 😄

jettieb commented on src/main/java/efub/assignment/community/member/domain/Member.java:
```diff
@@ -0,0 +1,90 @@
+    private Long memberId;
+
+    @Column(nullable = false, unique = true)
+    private int studentNumber;
```
사소하긴 한데, 학번은 주로 String 형으로 사용해요!
더하기/빼기 같은 연산을 하는 값이 아니기도 하고, 이대 학번의 경우에는 괜찮지만, 확장성 있게 생각하면 학번이 001234 인 경우에는 int 로 저장하면 1234 로 바뀌어버리는 문제 등으로 주로 String으로 사용합니다 😃

jettieb commented on src/main/java/efub/assignment/community/member/memberController.java:
```diff
@@ -0,0 +1,63 @@
+
+@RestController
+@RequestMapping("/members")
+public class memberController {
```
클래스 이름은 주로 대문자로 시작합니다! MemberController로 수정해주시는게 좋을 것 같아요~!! 😺

jettieb commented on src/main/java/efub/assignment/community/member/memberController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    //회원가입 API
+    @PostMapping("signup")
```
api 주소 앞에 /를 붙이는게 좀 더 명확합니다!
그리고 회원가입은 주로 POST `/members` 로 처리하는 경우가 많습니다!ㅎㅎ
/members/signup도 명확하지만, 그냥 /members로 추천드려요!

jettieb commented on src/main/java/efub/assignment/community/member/memberController.java:
```diff
@@ -0,0 +1,63 @@
+    // memberId에 해당하는 회원 정보 조회
+    @GetMapping("/{memberId}")
+    public MemberResponseDTO getMemberByMemberId(@PathVariable Long memberId){
+        return memberService.getMemberByMemberId(memberId);
```
응답 포맷으로 ResponseEntity로 일관되게 감싸주시는게 좋아요!
단순히 데이터를 반환하는 것 이상으로 응답 상태 코드나 헤더를 세밀하게 제어할 수 있어서 반환할때는 ResponseEntity에 감싸줍니다!

jettieb commented on src/main/java/efub/assignment/community/member/memberController.java:
```diff
@@ -0,0 +1,63 @@
+    public String changePassword(@PathVariable Long memberId,
+                               @RequestParam String newPassword){
+        memberService.changePassword(memberId, newPassword);
+        return "비밀번호가 성공적으로 변경되었습니다.";
```
단순 String으로 반환하면 간단하고, 응답 메시지를 빠르게 확인 가능하긴 하지만, 위와 같은 이유로 닉네임 변경, 비밀번호 변경, 회원 삭제 모두 ResponseEntity에 감싸서 반환해주는게 좋아요! :)

jettieb commented on src/main/resources/application.yml:
application.yml 에 민감정보인 db 비밀번호가 노출되고 있어서, 나중에 배포나 협업할 때 노출 위험이 있어서 .gitignore 처리하거나 환경 변수로 분리하는 걸 추천드립니다!

m2nsp committed:
♻️refactor: week4 - 코드리뷰 기반 리팩토링

- service, repository, controller 패키지 분리
- dto 반환 메소드 처리
- 조회 시 transactional 어노테이션 추가
- 학번 String 타입변환
- 클래스 이름 대문자 수정
- 응답 포맷 ResponseEntity로 통일
- configuration files .gitignore 추가

m2nsp committed:
💚chore: application.yml Git 추적 제거 및 .gitignore 추가
